src/Core/Material/BlinnPhongMaterialModel.cpp

```diff
@@ -24,6 +25,88 @@ void BlinnPhongMaterialModel::displayInfo() const {
     print( hasNormalTexture(), " Normal Texture : ", m_texNormal );
     print( hasOpacityTexture(), " Alpha Texture  : ", m_texOpacity );
 }
+Utils::Color Material::BlinnPhongMaterialModel::evalBSDF( Vector3 w_i,
+                                                          Vector3 w_o,
+                                                          Vector3 normal,
+                                                          Vector2 uv ) {
+    // diffuse lambertien component
+    Utils::Color diffuse = m_kd / M_PI;
+
+    // Blinn-Phong specular component
+    Vector3 halfway = w_i + w_o;
+    halfway.normalize();
+    Scalar specularIntensity =
+        ( m_ns + 2.0f ) / ( 2.0f * M_PI ) * std::pow( normal.dot( halfway ), m_ns );
+    Utils::Color specular = m_ks * specularIntensity;
+
+    // Combine the diffuse and specular components
+    Utils::Color bsdf = diffuse + specular;
+
+    return bsdf;
+}
+
+std::optional<std::pair<Vector3, Scalar>> BlinnPhongMaterialModel::sample( Vector3 inDir,
+                                                                           Vector3 normal,
+                                                                           Vector3 tangent,
+                                                                           Vector3 bitangent,
+                                                                           Vector2 u ) {
+    Vector3 halfway;
+
+    Vector3 rgbToLuminance { 0.2126_ra, 0.7152_ra, 0.0722_ra };
+    Scalar dIntensity   = m_kd.rgb().dot( rgbToLuminance );
+    Scalar sIntensity   = m_ks.rgb().dot( rgbToLuminance );
+    Scalar diffSpecNorm = std::max( 1_ra, dIntensity + sIntensity );
+
+    dIntensity /= diffSpecNorm;
+    sIntensity /= diffSpecNorm;
```

Comment:
These should be precomputed once (or each time the material parameter changes) so that you can sample several time the material without recomputing uniform expressions. This could be stored in a member structure representing  precomputed sampling parameters.